homeassistant/components/geocaching/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""The Geocaching integration."""
+import voluptuous as vol
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import ConfigEntry, ConfigEntryAuthFailed
+from homeassistant.const import CONF_CLIENT_ID, CONF_CLIENT_SECRET
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_entry_oauth2_flow import (
+    OAuth2Session,
+    async_get_config_entry_implementation,
+)
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.update_coordinator import UpdateFailed
+
+from .config_flow import GeocachingFlowHandler
+from .const import DOMAIN
+from .coordinator import GeocachingDataUpdateCoordinator
+from .oauth import GeocachingOAuth2Implementation
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_CLIENT_ID): cv.string,
+                vol.Required(CONF_CLIENT_SECRET): cv.string,
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+PLATFORMS = [SENSOR_DOMAIN]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType):
+    """Set up the Geocaching component."""
+    hass.data[DOMAIN] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN][CONF_CLIENT_ID] = config[DOMAIN][CONF_CLIENT_ID]
+
+    GeocachingFlowHandler.async_register_implementation(
+        hass,
+        GeocachingOAuth2Implementation(
+            hass,
+            client_id=config[DOMAIN][CONF_CLIENT_ID],
+            client_secret=config[DOMAIN][CONF_CLIENT_SECRET],
+            name="Geocaching",
+        ),
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Geocaching from a config entry."""
+    implementation = await async_get_config_entry_implementation(hass, entry)
+
+    oauth_session = OAuth2Session(hass, entry, implementation)
+    coordinator = GeocachingDataUpdateCoordinator(
+        hass, entry=entry, session=oauth_session
+    )
+
+    try:
+        await coordinator.async_refresh()
+    except UpdateFailed as exc:
+        raise ConfigEntryAuthFailed from exc
```

Comment:
Guess we need to update the scaffold :)

```suggestion
    await coordinator.async_config_entry_first_refresh()
```